044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,156 @@
+* The controller processes in each pod form a quorum
+* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to STARTED and are marked as ready
+* All pods are now ready
+
+However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:
+
+* The first pod is scheduled
+* The controller process starts up but cannot form a quorum because the other controllers are missing
+* The broker process starts up but since there is no controller quorum it does not move to STARTED, it stays in STARTING. This means the pod is not marked as ready
+* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff
+
+## Proposed order of tasks to complete this proposal
+
+Since changes to the KafkaRoller and support for non-combined mode is needed to fully implement this proposal, the following phases of development are proposed:
+
+### Phase 1 - Strimzi only supports combined mode
+* Combined mode pods are marked as both alive and ready when the pod is listening on 9090. 
+This is an improvement on the current state where the pod is marked as alive and ready as soon as it starts up.
+* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
+This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.
+
+### Phase 2 - Strimzi adds support for broker only and controller only modes
+* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
+* Broker only pods are marked as both alive and ready if listening on port 9091.
+* Combined mode pods are marked as both alive and ready when the pod is listening on 9090 (no change from phase 1).
+
+### Phase 3 - KafkaRoller is updated
+* KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods.
+
+### Phase 4
+* Broker only readiness probe fully implemented to match this proposal.
+* Combined mode readiness probe fully implemented to match this proposal.
+
+## Affected/not affected projects
+
+This change only affects the Cluster Operator and Kafka brokers.
+
+## Compatibility
+
+As part of this change the readiness probe for a Kafka broker in a ZooKeeper based cluster will be updated to 
+continue being marked as ready when the BrokerState metric > 3 && != 127.
+This means it will tolerate the PENDING_CONTROLLED_SHUTDOWN and SHUTTING_DOWN states.
+This will not cause any compatibility problems because currently the KafkaAgent stops running once the metric reaches 
+the RUNNING state.
+The change will make the code between KRaft and ZooKeeper mode simpler and protect readiness if in future the agent is 
+updated to continue running once the broker is "ready".
+
+## Rejected alternatives
```

Comment:
You probably don't need the header if you don' have any content.

Reply:
Good point, I've removed it